### PR TITLE
Oracle netplan v2 migration

### PR DESCRIPTION
**Pull request by a-dubs**
## Proposed Commit Message

```
feat(oracle): update Oracle Datasource to use netplan v2
```

## Test Steps

Ran the full integration test suite against Noble using a custom deb created from this feature branch.

## Merge type

- [x] Squash merge using "Proposed Commit Message"

### CODE REVIEW

File: cloudinit/sources/DataSourceOracle.py [hunk 1]
```diff
@@ -167,9 +174,16 @@ def _unpickle(self, ci_pkl_version: int) -> None:
             )
         if not hasattr(self, "_network_config"):
             self._network_config = {"config": [], "version": 1}
+        if not hasattr(self, "_network_config_v1"):
+            self._network_config_v1 = {"config": [], "version": 1}
 
     def _has_network_config(self) -> bool:
-        return bool(self._network_config.get("config", []))
+        if self._network_config:
```
Comment (TheRealFalcon) [inline cloudinit/sources/DataSourceOracle.py:181]: Nit: I think this whole section could be `return "config" in self._network_config or "ethernets" in self._network_config`

File: cloudinit/sources/DataSourceOracle.py [hunk 2]
```diff
@@ -149,7 +155,8 @@ def __init__(self, sys_cfg, *args, **kwargs):
             ]
         )
         self._network_config_source = KlibcOracleNetworkConfigSource()
-        self._network_config: dict = {"config": [], "version": 1}
+        self._network_config_v1: dict = {"config": [], "version": 1}
```
Comment (TheRealFalcon) [inline cloudinit/sources/DataSourceOracle.py:158]: I don't think we want to store a separate dictionary here. `self._network_config` should be able to store either v1 or v2.

File: cloudinit/sources/DataSourceOracle.py [hunk 3]
```diff
@@ -380,71 +398,148 @@ def _add_network_config_from_opc_imds(self, set_primary: bool = False):
...
+                "name": name,
+                "type": "physical",
+                "mac_address": mac_address,
+                "mtu": MTU,
+                "subnets": subnets,
+            }
+            self._network_config_v1["config"].append(interface_config)
+
+            ############################# V2 HERE #############################
+            # Why does this elif exist???
```
Comment (TheRealFalcon) [inline cloudinit/sources/DataSourceOracle.py:440]: We can get rid of my snarky comments :rofl:

File: cloudinit/sources/DataSourceOracle.py [hunk 4]
```diff
@@ -380,71 +398,148 @@ def _add_network_config_from_opc_imds(self, set_primary: bool = False):
...
+                if vnic_dict.get("privateIp"):
+                    interface_config["addresses"].append(
+                        f"{vnic_dict['privateIp']}/{network.prefixlen}"
+                    )
+                if vnic_dict.get("ipv6Addresses"):
+                    interface_config["addresses"].append(
+                        f"{vnic_dict['ipv6Addresses'][0]}/"
+                        f"{network.prefixlen}"
+                    )
+            self._network_config["ethernets"][name] = interface_config
```
Comment (TheRealFalcon) [inline cloudinit/sources/DataSourceOracle.py:465]: I think everything up until here looks generally good. I haven't checked specific keys yet, but the general idea makes sense.

I think everything after here would make more sense in tests. It's good to make sure that our v2 rendering matches (or closely resembles :smile: ) our v1 rendering as you're doing, but we don't need to be converting or comparing them in production code.

Instead, I think it'd be most useful to have tests containing a few different realistic examples of networking data we get from Oracle, then take the rendered v2 generated here and compare that against  the rendered v1 from here, and feeding that through netplan to get a v2 to compare against our rendered v2 here.

**Review by TheRealFalcon — COMMENTED** [review]
See my inline comments. I think you added some good code, but much of it can move into tests.

Additionally, I forget, why are we moving to support both v1 and v2 now? Once we move to v2, is there any reason we would need to keep the v1 support around?

**Comment by a-dubs** [comment]
@TheRealFalcon thanks for the feedback! I guess the idea here was that by keeping the old v1 code around, we could do the comparisons on live systems that way if anyone ever finds a weird edge case then it will log a warning and they could file a bug. But honestly, now that you mention it, I do kinda prefer the idea of just having a super comprehensive unit test suite to verify it once and be done and then move on with our lives. 

@holmanb thoughts? since this was your idea originally?

**Comment by holmanb** [comment]
> @TheRealFalcon thanks for the feedback! I guess the idea here was that by keeping the old v1 code around, we could do the comparisons on live systems that way if anyone ever finds a weird edge case then it will log a warning and they could file a bug. But honestly, now that you mention it, I do kinda prefer the idea of just having a super comprehensive unit test suite to verify it once and be done and then move on with our lives.
> 
> @holmanb thoughts? since this was your idea originally?

The only benefit that I see of a runtime check is that it might catch edge cases that couldn't be predicted with a battery of unit tests. I'm fine with just getting some good test coverage instead if that's preferred.

File: cloudinit/sources/DataSourceOracle.py (hunk 3 above)
Comment (a-dubs) [reply cloudinit/sources/DataSourceOracle.py:440]: XD will do bahahahah

File: cloudinit/sources/DataSourceOracle.py (hunk 2 above)
Comment (a-dubs) [reply cloudinit/sources/DataSourceOracle.py:158]: this was for the purpose of comparing v1 vs v2 at runtime, but now that we are moving away from that approach, this can be removed.

File: tests/unittests/sources/test_oracle.py [hunk 5]
```diff
@@ -1794,3 +1704,288 @@ def test_get_versioned_metadata_base_url(
...
+        if not config_b:
+            raise ValueError("config_b or rendered_string_b must be provided")
+        network_state_b = network_state.parse_net_config_data(
+            config_b, renderer=netplan.Renderer  # type: ignore
+        )
+        rendered_string_b = netplan.Renderer()._render_content(
+            network_state_b,
+        )
+
+    diff = DeepDiff(
```
Comment (holmanb) [inline tests/unittests/sources/test_oracle.py:1737]: There is only one call to DeepDiff, so I would discourage an external dependency - especially given that the inputs are known.

Would a generic recursive function that sorts lists within each dict would suffice?

**Review by holmanb — COMMENTED** [review]
@a-dubs I left some comments below.

File: tests/unittests/sources/test_oracle.py (hunk 5 above)
Comment (a-dubs) [reply tests/unittests/sources/test_oracle.py:1737]: agreed this would be sufficient and better than pulling in a new dependency.

File: cloudinit/sources/DataSourceOracle.py [hunk 6]
```diff
@@ -380,75 +392,54 @@ def _add_network_config_from_opc_imds(self, set_primary: bool = False):
...
-        return False
-    return metadata_address.startswith(IPV4_METADATA_ROOT.split("opc")[0])
+def convert_v1_netplan_to_v2(network_config: dict) -> dict:
+    has_base_network_key = "network" in network_config
+    if has_base_network_key:
+        network_config = network_config["network"]
+    v1_network_state = network_state.parse_net_config_data(
+        network_config, renderer=netplan.Renderer  # type: ignore
+    )
+    netplan_v2_yaml_string = netplan.Renderer()._render_content(
```
Comment (TheRealFalcon) [inline cloudinit/sources/DataSourceOracle.py:432]: I don't actually think this is safe long term. Since v2 is a subset of netplan, it's possible that we'll render something that is valid netplan but not valid v2. I'm wondering if we might just need a separate klibc->v2 renderer rather than trying to support all of v1->v2.

**Review by TheRealFalcon — CHANGES_REQUESTED** [review]
I left some inline comments.

**Comment by a-dubs** [comment]
This is being closed in favor of the new approach in #6097
